src/titiler/multidim/reader.py

```diff
@@ -1,24 +1,208 @@
+def _is_dir(store, path: str = "") -> bool:
+    """Return True if path is a prefix containing any objects (directory-like)."""
+    # sanitize path and slashes
+    path = path.rstrip("/") + "/"
+    stream = store.list(prefix=path, chunk_size=1)
+    try:
+        batch = next(stream)
+        return len(batch) > 0
+    except StopIteration:
+        return False
+
+
+def identify_storage_backend(src_path: str) -> str:
+    """Identify the storage backend for a given path."""
+    parsed = urlparse(src_path)
+    protocol = parsed.scheme or "file"
+
+    if protocol == "file":
+        store = obstore.store.LocalStore(src_path)
+    elif protocol == "s3":
+        store = obstore.store.S3Store(
+            bucket=parsed.netloc,
+            prefix=parsed.path.lstrip("/"),
+        )
+    else:
+        raise NotImplementedError(
+            f"Storage backend identification for protocol {protocol} is not implemented"
+        )
+
+    is_dir = _is_dir(store)
+    if not is_dir:
+        # assume this is a file, and detect the format based on the file extension
+        _, ext = os.path.splitext(parsed.path)
+        if ext in [".nc", ".nc4"]:
+            format = "h5netcdf"
+        else:
+            raise NotImplementedError(
+                f"File format identification for extension {ext} is not implemented"
+            )
+    else:
+        has_manifests = _is_dir(store, "manifests")
+        if has_manifests:
+            format = "icechunk"
+        else:
+            format = "zarr"
+    return format
+
+
+def guess_opener(
+    src_path: str,
+    group: Optional[Any] = None,
+    decode_times: bool = True,
+    authorize_virtual_chunk_access: Optional[Dict[str, Dict[str, Any]]] = None,
+    **kwargs: Any,
+) -> xr.Dataset:
+    """Guess the storage backend and return an xarray Dataset.
+
+    Args:
+        src_path: Path to the dataset
+        group: Optional group/subgroup to open
+        decode_times: Whether to decode time coordinates
+        authorize_virtual_chunk_access: Authorization config for icechunk virtual chunks
+        **kwargs: Additional arguments to pass to the opener
+
+    Returns:
+        xarray.Dataset
+    """
+
+    # Identify the storage backend
+    storage_format = identify_storage_backend(src_path)
+
+    if storage_format == "icechunk":
+        return opener_icechunk(
+            src_path,
+            group=group,
+            decode_times=decode_times,
+            authorize_virtual_chunk_access=authorize_virtual_chunk_access,
+        )
+    else:
+        # For zarr, h5netcdf, or other formats, use the standard xarray opener
+        return xarray_open_dataset(
+            src_path, group=group, decode_times=decode_times, **kwargs
+        )
+
+
+def _inject_settings(
+    options: Dict[str, Any], settings: BaseSettings, field: str
+) -> Dict[str, Any]:
+    if field not in options:
+        options[field] = api_settings.authorized_chunk_access
+    return options
```

Comment:
I added this little helper, just to reduce the duplicated code. Having less code in `list_variables` also helps me accept the repeated call to `guess_opener` hehe.